Exception is thrown when source (e.g. LDAP) password changes within the cache TTL window

**Problem**
We are experiencing INTERNAL_ERROR when trying to login with old password within cache_ttl timeout when rabbitmq_auth_backend_cache is used.

**Use-case**
1) Open AMQP connection successfully as user1, authentication done via LDAP with rabbitmq_auth_backend_cache.
2) Close AMQP connection successfully as user1.
3) Change password in ldap for user1.
4) Try to open AMQP connection as user1 with old password within cache_ttl. -> Java client response: Caused by: com.rabbitmq.client.ShutdownSignalException: connection error; protocol method: #method<connection.close>(reply-code=541, reply-text=INTERNAL_ERROR, class-id=0, method-id=0).

**Note**
This use-case works as expected without the rabbitmq_auth_backend_cache.

**Versions**
Erlang version 22, RMQ version 3.8.0,

**Plugins**
`[rabbitmq_auth_backend_ldap,rabbitmq_auth_backend_cache,rabbitmq_management,rabbitmq_event_exchange,rabbitmq_tracing,rabbitmq_message_timestamp].`

**RMQ config**
```
[
  {kernel,
    [
      {inet_default_connect_options, [{nodelay, true}]},
      {inet_default_listen_options, [{nodelay, true}]},
      {net_ticktime, 120}
    ]},
  {rabbit,
    [
      {log_levels, [{connection, debug}, {channel, debug}]},
      {cluster_partition_handling, autoheal},
      {cluster_keepalive_interval, 10000},
      {disk_free_limit, 100000000},
      {loopback_users, []},
      {channel_max, 0},
      {hipe_compile, false},
      {collect_statistics_interval, 10000},
      {tcp_listen_options, [{backlog, 4096}, {nodelay, true}]},
      {auth_backends,
        [
          rabbit_auth_backend_internal,
          rabbit_auth_backend_cache
        ]
      }
    ]},

  {rabbitmq_tracing,
    [
      {directory, "/var/log/rabbitmq/tracing"}
    ]
  },
  {rabbitmq_management,
    [
      {load_definitions, "/etc/rabbitmq/rabbitmq-definitions.json"},
      {redirect_old_port, false}
    ]},
  {rabbitmq_auth_backend_cache,
    [
      {cached_backend, rabbit_auth_backend_ldap},
      {cache_ttl, 50000}
    ]},
  {rabbitmq_event_exchange,
    [
      {vhost, <<"/test">>}
    ]},
  {rabbitmq_auth_backend_ldap,
    [
      {servers, ["lts-ldap"]},
      {port, 389},
      {user_dn_pattern, "uid=${username},dc=test"},
      {use_ssl, false},
      {timeout, 5000},
      {log, "network_unsafe"},
      {vhost_access_query,
        {equals, {string, "${vhost}"}, {string, "/test"}}
      }
    ]
  }
].
```

**Stack trace**
```
2019-10-29 09:30:35 =ERROR REPORT====
** Generic server <0.2539.0> terminating
** Last message in was {'$gen_cast',{method,{'queue.declare',0,<<"xxx.user1.9efcdbe1-152d-47cb-9759-387fb6d96aa9">>,false,false,true,true,false,[]},none,noflow}}
** When Server state == {ch,{conf,running,rabbit_framing_amqp_0_9_1,2,<0.2527.0>,<0.2537.0>,<0.2527.0>,<<"172.21.0.1:41524 -> 172.21.0.8:5672">>,rabbit_reader,{user,<<"user1">>,[],[{rabbit_auth_backend_cache,{impl,"uid=user1,dc=test",<<"a">>}}]},<<"/test">>,<<>>,<0.2528.0>,[{<<"exchange_exchange_bindings">>,bool,true},{<<"connection.blocked">>,bool,true},{<<"authentication_failure_close">>,bool,true},{<<"basic.nack">>,bool,true},{<<"publisher_confirms">>,bool,true},{<<"consumer_cancel_notify">>,bool,true}],none,0,134217728,undefined},{lstate,<0.2538.0>,false},none,1,{0,{[],[]}},#{},{state,#{},erlang},#{},#{},{set,0,16,16,8,80,48,{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]},{{[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]}}},{state,fine,10000,#Ref<0.708702737.2879389697.142542>},false,1,{unconfirmed,{0,nil},#{},#{}},[],[],none,flow,[],#{},#Ref<0.708702737.2879389697.142536>}
** Reason for termination ==
** {{case_clause,{refused,"uid=user1,dc=test",[]}},[{rabbit_access_control,check_access,5,[{file,"src/rabbit_access_control.erl"},{line,211}]},{lists,foldl,3,[{file,"lists.erl"},{line,1263}]},{rabbit_channel,check_resource_access,4,[{file,"src/rabbit_channel.erl"},{line,1006}]},{rabbit_channel,handle_method,6,[{file,"src/rabbit_channel.erl"},{line,2531}]},{rabbit_channel,handle_method,3,[{file,"src/rabbit_channel.erl"},{line,1661}]},{rabbit_channel,handle_cast,2,[{file,"src/rabbit_channel.erl"},{line,627}]},{gen_server2,handle_msg,2,[{file,"src/gen_server2.erl"},{line,1067}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,249}]}]}
2019-10-29 09:30:35 =CRASH REPORT====
  crasher:
    initial call: rabbit_channel:init/1
    pid: <0.2539.0>
    registered_name: []
    exception exit: {{{case_clause,{refused,"uid=user1,dc=test",[]}},[{rabbit_access_control,check_access,5,[{file,"src/rabbit_access_control.erl"},{line,211}]},{lists,foldl,3,[{file,"lists.erl"},{line,1263}]},{rabbit_channel,check_resource_access,4,[{file,"src/rabbit_channel.erl"},{line,1006}]},{rabbit_channel,handle_method,6,[{file,"src/rabbit_channel.erl"},{line,2531}]},{rabbit_channel,handle_method,3,[{file,"src/rabbit_channel.erl"},{line,1661}]},{rabbit_channel,handle_cast,2,[{file,"src/rabbit_channel.erl"},{line,627}]},{gen_server2,handle_msg,2,[{file,"src/gen_server2.erl"},{line,1067}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,249}]}]},[{gen_server2,terminate,3,[{file,"src/gen_server2.erl"},{line,1183}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,249}]}]}
    ancestors: [<0.2536.0>,<0.2531.0>,<0.2526.0>,<0.2525.0>,<0.894.0>,<0.893.0>,<0.892.0>,rabbit_sup,<0.272.0>]
    message_queue_len: 0
    messages: []
    links: [<0.2536.0>]
    dictionary: [{rand_seed,{#{jump => #Fun<rand.3.8986388>,max => 288230376151711743,next => #Fun<rand.2.8986388>,type => exsplus},[66234829244152337|5836671951412012]}},{channel_operation_timeout,15000},{process_name,{rabbit_channel,{<<"172.21.0.1:41524 -> 172.21.0.8:5672">>,2}}},{permission_cache_can_expire,false}]
    trap_exit: true
    status: running
    heap_size: 4185
    stack_size: 27
    reductions: 6992
  neighbours:
2019-10-29 09:30:35 =SUPERVISOR REPORT====
     Supervisor: {<0.2536.0>,rabbit_channel_sup}
     Context:    child_terminated
     Reason:     {{case_clause,{refused,"uid=user1,dc=test",[]}},[{rabbit_access_control,check_access,5,[{file,"src/rabbit_access_control.erl"},{line,211}]},{lists,foldl,3,[{file,"lists.erl"},{line,1263}]},{rabbit_channel,check_resource_access,4,[{file,"src/rabbit_channel.erl"},{line,1006}]},{rabbit_channel,handle_method,6,[{file,"src/rabbit_channel.erl"},{line,2531}]},{rabbit_channel,handle_method,3,[{file,"src/rabbit_channel.erl"},{line,1661}]},{rabbit_channel,handle_cast,2,[{file,"src/rabbit_channel.erl"},{line,627}]},{gen_server2,handle_msg,2,[{file,"src/gen_server2.erl"},{line,1067}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,249}]}]}
     Offender:   [{pid,<0.2539.0>},{name,channel},{mfargs,{rabbit_channel,start_link,[2,<0.2527.0>,<0.2537.0>,<0.2527.0>,<<"172.21.0.1:41524 -> 172.21.0.8:5672">>,rabbit_framing_amqp_0_9_1,{user,<<"user1">>,[],[{rabbit_auth_backend_cache,{impl,"uid=user1,dc=test",<<"a">>}}]},<<"/test">>,[{<<"exchange_exchange_bindings">>,bool,true},{<<"connection.blocked">>,bool,true},{<<"authentication_failure_close">>,bool,true},{<<"basic.nack">>,bool,true},{<<"publisher_confirms">>,bool,true},{<<"consumer_cancel_notify">>,bool,true}],<0.2528.0>,<0.2538.0>]}},{restart_type,intrinsic},{shutdown,70000},{child_type,worker}]

2019-10-29 09:30:35 =SUPERVISOR REPORT====
     Supervisor: {<0.2536.0>,rabbit_channel_sup}
     Context:    shutdown
     Reason:     reached_max_restart_intensity
     Offender:   [{pid,<0.2539.0>},{name,channel},{mfargs,{rabbit_channel,start_link,[2,<0.2527.0>,<0.2537.0>,<0.2527.0>,<<"172.21.0.1:41524 -> 172.21.0.8:5672">>,rabbit_framing_amqp_0_9_1,{user,<<"user1">>,[],[{rabbit_auth_backend_cache,{impl,"uid=user1,dc=test",<<"a">>}}]},<<"/test">>,[{<<"exchange_exchange_bindings">>,bool,true},{<<"connection.blocked">>,bool,true},{<<"authentication_failure_close">>,bool,true},{<<"basic.nack">>,bool,true},{<<"publisher_confirms">>,bool,true},{<<"consumer_cancel_notify">>,bool,true}],<0.2528.0>,<0.2538.0>]}},{restart_type,intrinsic},{shutdown,70000},{child_type,worker}]
```

--- a/src/rabbit_auth_backend_ldap.erl
+++ b/src/rabbit_auth_backend_ldap.erl
@@ -105,10 +105,12 @@ check_vhost_access(User = #auth_user{username = Username,
             {user_dn,  UserDN},
             {vhost,    VHost}] ++ OptionsArgs ++ ADArgs,
     ?L("CHECK: ~s for ~s", [log_vhost(Args), log_user(User)]),
-    R = evaluate_ldap(env(vhost_access_query), Args, User),
-    ?L("DECISION: ~s for ~s: ~p",
-       [log_vhost(Args), log_user(User), log_result(R)]),
-    R.
+    R0 = evaluate_ldap(env(vhost_access_query), Args, User),
+    R1 = ensure_rabbit_authz_backend_result(R0),
+    ?L("DECISION: ~s for ~s: ~p (~p)",
+       [log_vhost(Args), log_user(User),
+        log_result(R0), log_result(R1)]),
+    R1.
 
 check_resource_access(User = #auth_user{username = Username,
                                         impl     = #impl{user_dn = UserDN}},
@@ -124,10 +126,12 @@ check_resource_access(User = #auth_user{username = Username,
             {name,       Name},
             {permission, Permission}] ++ OptionsArgs ++ ADArgs,
     ?L("CHECK: ~s for ~s", [log_resource(Args), log_user(User)]),
-    R = evaluate_ldap(env(resource_access_query), Args, User),
-    ?L("DECISION: ~s for ~s: ~p",
-       [log_resource(Args), log_user(User), log_result(R)]),
-    R.
+    R0 = evaluate_ldap(env(resource_access_query), Args, User),
+    R1 = ensure_rabbit_authz_backend_result(R0),
+    ?L("DECISION: ~s for ~s: ~p (~p)",
+       [log_resource(Args), log_user(User),
+        log_result(R0), log_result(R1)]),
+    R1.
 
 check_topic_access(User = #auth_user{username = Username,
                                      impl     = #impl{user_dn = UserDN}},
@@ -143,14 +147,30 @@ check_topic_access(User = #auth_user{username = Username,
             {name,       Name},
             {permission, Permission}] ++ OptionsArgs ++ ADArgs,
     ?L("CHECK: ~s for ~s", [log_resource(Args), log_user(User)]),
-    R = evaluate_ldap(env(topic_access_query), Args, User),
-    ?L("DECISION: ~s for ~s: ~p",
-        [log_resource(Args), log_user(User), log_result(R)]),
-    R.
+    R0 = evaluate_ldap(env(topic_access_query), Args, User),
+    R1 = ensure_rabbit_authz_backend_result(R0),
+    ?L("DECISION: ~s for ~s: ~p (~p)",
+        [log_resource(Args), log_user(User),
+         log_result(R0), log_result(R1)]),
+    R1.
 
 state_can_expire() -> false.
 
 %%--------------------------------------------------------------------
+
+ensure_rabbit_authz_backend_result(true) ->
+    true;
+ensure_rabbit_authz_backend_result(false) ->
+    false;
+ensure_rabbit_authz_backend_result({error, _}=Error) ->
+    Error;
+% rabbitmq/rabbitmq-auth-backend-ldap#116
+ensure_rabbit_authz_backend_result({refused, _, _}) ->
+    false;
+ensure_rabbit_authz_backend_result({ok, _}) ->
+    true;
+ensure_rabbit_authz_backend_result({ok, _, _}) ->
+    true.
 
 context_as_options(Context, Namespace) when is_map(Context) ->
     % filter keys that would erase fixed variables

